src/exception/exception-api.ts

```diff
@@ -17,87 +17,134 @@
-import { RepoError } from './exception-repo';
+import { ExceptionTSVError } from './error-handling';
 import * as exceptionService from './exception-service';
-import { isProgramExceptionRecord, isReadonlyArrayOf } from './types';
+import programExceptionRepository from './repo/program';
+import {
+  EntityExceptionRecord,
+  isEntityExceptionRecord,
+  isProgramExceptionRecord,
+  isReadonlyArrayOf,
+  ProgramExceptionRecord,
+} from './types';
 
 const L = loggerFor(__filename);
 
-const ProgramExceptionErrorMessage = {
-  TSV_PARSING_FAILED: `This file is formatted incorrectly`,
-} as const;
-
-function getResStatus(result: exceptionService.Result): number {
-  if (result.success) {
-    return 200;
-  } else if (result.error?.code === RepoError.DOCUMENT_UNDEFINED) {
-    return 400;
-  } else {
-    return 500;
+type ValidateRecords<T> = (records: ReadonlyArray<TsvUtils.TsvRecordAsJsonObj>) => ReadonlyArray<T>;
+
+const validateProgramExceptionRecords: ValidateRecords<ProgramExceptionRecord> = records => {
+  if (!isReadonlyArrayOf(records, isProgramExceptionRecord)) {
+    L.debug(`Program Exception TSV_PARSING_FAILED`);
+    throw new ExceptionTSVError('Invalid program exception tsv file');
   }
-}
+  return records;
+};
+
+const validateEntityExceptionRecords: ValidateRecords<EntityExceptionRecord> = records => {
+  if (!isReadonlyArrayOf(records, isEntityExceptionRecord)) {
+    L.debug(`Entity Exception TSV_PARSING_FAILED`);
+    throw new ExceptionTSVError('Invalid entity exception tsv file');
+  }
+  return records;
+};
 
 class ExceptionController {
   @HasFullWriteAccess()
-  async createProgramException(req: Request, res: Response) {
-    if (!requestContainsFile(req, res)) {
-      return false;
-    }
 
+  // program exceptions
+  async createProgramException(req: Request, res: Response) {
     const programId = req.params.programId;
-    const file = req.file;
-
-    try {
-      const records = await TsvUtils.tsvToJson(file.path);
-      if (records.length === 0) {
-        throw new Error('TSV has no records!');
-      }
-
-      if (!isReadonlyArrayOf(records, isProgramExceptionRecord)) {
-        throw new Error('TSV is incorrectly structured');
-      }
-
-      const result = await exceptionService.operations.createProgramException({
-        programId,
-        records,
-      });
-
-      if (!result.success) {
-        return res.status(422).send(result);
-      }
-      return res.status(201).send(result);
-    } catch (err) {
-      L.error(`Program Exception TSV_PARSING_FAILED`, err);
-      return ControllerUtils.unableToProcess(res, ProgramExceptionErrorMessage.TSV_PARSING_FAILED);
-    }
+    const records = await parseTSV(req.file.path);
+    const programExceptionRecords = validateProgramExceptionRecords(records);
+
+    const result = await exceptionService.operations.createProgramException({
+      programId,
+      records: programExceptionRecords,
+    });
+
+    const status = !result.success ? 422 : 201;
```

Comment:
As personal policy, I wouldn’t not ask not to use unnecessary negative conditions 😂
```suggestion
    const status = result.success ? 201 : 422;
```